File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode>pq = new PriorityQueue<>((a,b)->a.val-b.val);
        for(ListNode node:lists) {
            if(node !=null)
            pq.add(node);
    }
        ListNode ans = new ListNode(0);
        ListNode prev = ans;
        while(!pq.isEmpty()){
            ListNode temp = pq.poll();
            ans.next = new ListNode(temp.val);
            ans=ans.next;
            if(temp.next!=null) pq.add(temp.next);
        }
        return prev.next;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        //use minheap to sort in asc 
        auto compare = [](ListNode* a, ListNode* b) {
            return a->val > b->val;  // Min-heap (ascending order)
        };

        // Min-heap to store ListNode* based on node values
        priority_queue<ListNode*, vector<ListNode*>, decltype(compare)> pq(compare);

        for(auto l:lists){
            if(l!=NULL) pq.push(l);
        }
        ListNode* head=new ListNode(0);
        ListNode* tail=head;
        while(!pq.empty()){
            ListNode* nextL=pq.top();
            pq.pop();
            tail->next= nextL;
            tail=tail->next;
            if(tail->next!=NULL) pq.push(tail->next);
            
        }
        return head->next;
    }
};